docs/DifferentiableTypes.md

```diff
@@ -0,0 +1,281 @@
+
+## Design overview
+
+The `Differentiable` protocol generalizes all types that can be used as arguments and results of differentiable functions.
+
+The compiler can automatically provide default implementations of `Differentiable` protocol requirements for struct types whose stored properties all conform to `Differentiable`.
+
+Here are some examples:
+
+```swift
+struct Vector: Differentiable, VectorNumeric {
+    // The compiler synthesizes all `Differentiable` protocol requirements
+    // when all stored properties conform to `Differentiable`.
+    var x, y, z: Float
+}
+
+// Differential operators like `gradient(at:in:)` just work!
+let v = Vector(x: 1, y: 2, z: 3)
+let 𝛁v = gradient(at: v) { v in (v + v).x }
+
+print(𝛁v)
+// Vector(x: 2.0, y: 0.0, z: 0.0)
+```
+
+A `Differentiable`-conforming type may have stored properties that are not meant to have a derivative with respect to `self`. Use the `@noDerivative` attribute to mark those properties; they will not have a corresponding entry in the synthesized `TangentVector`, `CotangentVector`, and `AllDifferentiableVariables` struct types.
+
+Here’s an example deep learning layer with some `@noDerivative` properties:
+
+```swift
+struct DenseLayer : Differentiable {
+    // These properties should have derivative values.
+    var weight: Tensor<Float>
+    var bias: Tensor<Float>
+
+    // These auxiliary properties should not have derivative values.
+    // Thus, they are marked with `@noDerivative`.
+    //
+    // `@noDerivative` properties do not have a corresponding entry
+    // in synthesized associated struct types.
+    @noDerivative var useBias: Bool = true
+    @noDerivative var previousWeight: Tensor<Float> = Tensor(0)
+
+    init(weight: Tensor<Float>, bias: Tensor<Float>) {
+        self.weight = weight
+        self.bias = bias
+    }
+
+    // The compiler synthesizes all `Differentiable` protocol requirements,
+    // adding only properties not marked with `@noDerivative` to associated
+    // tangent space types.
+
+    func applied(to input: Tensor<Float>) -> Tensor<Float> {
+        return matmul(input, weight) + bias
+    }
+}
+
+// Differential operators like `gradient(at:in:)` just work!
+let dense = DenseLayer(weight: [[1, 1], [1, 1]], bias: [0, 0])
+let 𝛁dense = gradient(at: dense) { dense in dense.applied(to: [[3, 3]]).sum() }
+
+dump(𝛁dense)
+// ▿ DenseLayer.AllDifferentiableVariables
+//   - weight: [[3.0, 3.0], [3.0, 3.0]]
+//   - bias: [1.0, 1.0]
+```
+
+## Protocol details
+
+Here is the full `Differentiable` protocol definition. More explanation is provided below.
+
+```swift
+/// A type that mathematically represents a differentiable manifold whose
+/// tangent spaces are finite-dimensional.
+public protocol Differentiable {
+    /// The tangent bundle of this differentiable manifold.
+    associatedtype TangentVector : AdditiveArithmetic & Differentiable
```

Comment:
```suggestion
    associatedtype TangentVector: AdditiveArithmetic & Differentiable
```
Same elsewhere.

Reply:
I noticed most of `AutoDiff.swift` in the Swift repo uses this syntax. I'm guessing there is a minor TODO JIRA ticket to update the file?

Reply:
The story is rather complicated. The standard library uses "the swift.org code example style", which is not officially documented and used only within the compiler repo. All our documentation and library code (like the deep learning library) use the [Google Swift Style](https://google.github.io/swift/) with the exception that indentations are 4 spaces long.

Reply:
Ahh, I see makes sense thanks!